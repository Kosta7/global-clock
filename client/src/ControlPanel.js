import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { GoogleLogin, GoogleLogout } from "react-google-login";

import { Popup } from "ui/Popup";

export class ControlPanel extends React.Component {
  state = {
    isLoginMode: false,
    isLoggedIn: false
  };

  componentDidMount() {}

  render() {
    return (
      <StyledControlPanel>
        <Button onClick={this.props.resetShift}>
          <span role="img" aria-label="reset">
            üîÉ
          </span>
        </Button>
        <Button
          onClick={this.props.toggleEditMode}
          isEditMode={this.props.isEditMode}
        >
          <span role="img" aria-label="edit">
            ‚úèÔ∏è
          </span>
        </Button>
        <Button onClick={this.openLoginConfirmation.bind(this)}>
          {this.state.isLoggedIn ? "logout" : "login"}
        </Button>
        {this.state.isLoginMode &&
          (this.state.isLoggedIn ? (
            <Popup onClose={this.closeLoginConfirmation.bind(this)}>
              <GoogleLogout
                buttonText="Logout"
                onLogoutSuccess={this.onLogoutSuccess.bind(this)}
              />
            </Popup>
          ) : (
            <Popup onClose={this.closeLoginConfirmation.bind(this)}>
              <GoogleLogin
                clientId={process.env.REACT_APP_GOOGLE_CLIENT_ID}
                buttonText="Login"
                onSuccess={this.onLoginSuccess.bind(this)}
                onFailure={this.onLoginFailure.bind(this)}
              />
            </Popup>
          ))}
      </StyledControlPanel>
    );
  }

  openLoginConfirmation() {
    this.setState({ isLoginMode: true });
  }

  closeLoginConfirmation() {
    this.setState({ isLoginMode: false });
  }

  onLoginSuccess(googleUser) {
    console.log(googleUser);
    const { id_token } = googleUser.getAuthResponse();
    const requestBody = { idToken: id_token };
    fetch("/tokensignin", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(requestBody)
    }).then(response => {
      if (response.ok) {
        this.props.updateClockList();
      }
    });
  }

  onLoginFailure(response) {
    console.log(response);
  }

  onLogoutSuccess(response) {
    console.log(response);
  }
}
ControlPanel.propTypes = {
  isEditMode: PropTypes.bool.isRequired,
  toggleEditMode: PropTypes.func.isRequired,
  resetShift: PropTypes.func.isRequired,
  updateClockList: PropTypes.func.isRequired
};

const Button = styled.button`
  flex-basis: 50%;
  border: 1px solid ${props => (props.isEditMode ? "#f5af5f" : "transparent")};
  background: ${props => (props.isEditMode ? "#ffe187" : "none")};
  font-size: 20px;
  font-family: monospace;
  margin: 2px;
  outline: none;

  &:active {
    border-color: #f5af5f;
    background-color: #ffe187;
  }

  &:disabled {
    opacity: 0.4;
  }
`;

const StyledControlPanel = styled.div`
  position: fixed;
  z-index: 2;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 50px;
  background-color: white;
  display: flex;
`;
