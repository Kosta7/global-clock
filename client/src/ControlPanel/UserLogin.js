import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import load from "load-script";

import { Popup } from "ui/Popup";

export class UserLogin extends React.Component {
  state = {
    isGapiLoaded: !!window.gapi
  };

  componentDidMount() {
    if (this.setState.isGapiLoaded) {
      this.renderLoginButton();
    } else {
      this.loadGapi(this.renderLoginButton);
    }
  }

  render() {
    return (
      <Popup onClose={this.props.closeUserPopup.bind(this)}>
        <StyledUserLogin>
          <Appeal>
            Log in with your Google account to be able to save clocks:
          </Appeal>
          <div id="login-button" />
        </StyledUserLogin>
      </Popup>
    );
  }

  loadGapi(callback) {
    load("https://apis.google.com/js/platform.js", (err, script) => {
      window.gapi.load("auth2", () => {
        window.gapi.auth2.init();
        callback.call(this);
      });
    });
  }

  renderLoginButton() {
    window.gapi.signin2.render("login-button", {
      scope: "openid",
      width: 120,
      height: 50,
      text: "Login",
      onsuccess: this.onSuccess.bind(this),
      onfailure: this.onFailure.bind(this)
    });
  }

  onSuccess(googleUser) {
    this.props.closeUserPopup();

    const idToken = googleUser.getAuthResponse().id_token;
    this.createSession(idToken);
  }

  onFailure(response) {
    console.error(response);
  }

  async createSession(idToken) {
    const response = await fetch("/auth/tokensignin", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ idToken })
    });
    if (response.ok) {
      this.props.updateClockList();
    }
  }
}
UserLogin.propTypes = {
  updateClockList: PropTypes.func.isRequired,
  closeUserPopup: PropTypes.func.isRequired
};

const Appeal = styled.div`
  color: #343434;
  font-size: 16px;
  margin-bottom: 10px;
`;

const StyledUserLogin = styled.div`
  padding: 10px;
`;
