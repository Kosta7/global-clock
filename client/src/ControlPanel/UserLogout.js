import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import load from "load-script";

import { Popup } from "ui/Popup";

export class UserLogout extends React.Component {
  state = {
    isGapiLoaded: !!window.gapi
  };

  componentDidMount() {
    if (!this.state.isGapiLoaded) {
      this.loadGapi();
    }
  }

  render() {
    return (
      <Popup onClose={this.props.closeUserPopup.bind(this)}>
        <StyledUserLogout>
          <Appeal>Log out of your Google account:</Appeal>
          <Button
            disabled={!this.state.isGapiLoaded}
            onClick={this.logout.bind(this)}
          >
            Logout
          </Button>
        </StyledUserLogout>
      </Popup>
    );
  }

  loadGapi() {
    load("https://apis.google.com/js/platform.js", (err, script) => {
      window.gapi.load("auth2", () => {
        window.gapi.auth2.init();
        this.setState({ isGapiLoaded: true });
      });
    });
  }

  logout() {
    document.cookie =
      "MyGoogleID=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;";
    const auth2 = window.gapi.auth2.getAuthInstance();
    auth2
      .signOut()
      .then(this.onSuccess.bind(this))
      .catch(this.onFailure.bind(this));
  }

  async onSuccess(googleUser) {
    this.props.closeUserPopup();
    fetch("/auth/tokensignout").then(response => {
      if (response.ok) {
        this.props.updateClockList();
      }
    });
    localStorage.userName = undefined;
  }

  onFailure(response) {
    console.error(response);
  }
}
UserLogout.propTypes = {
  updateClockList: PropTypes.func.isRequired,
  closeUserPopup: PropTypes.func.isRequired
};

const Button = styled.button`
  width: 120px;
  height: 50px;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.25);
  color: #757575;
  border-radius: 1px;
  border: 1px solid white;
  font-family: arial;
  font-size: 16px;
  background-color: #fff;
  cursor: pointer;
`;

const Appeal = styled.div`
  color: #343434;
  font-size: 16px;
  margin-bottom: 10px;
`;

const StyledUserLogout = styled.div`
  padding: 10px;
`;
