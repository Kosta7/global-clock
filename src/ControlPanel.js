import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

const ControlPanel = ({ editMode, toggleEditMode, resetShift, className }) => (
  <div className={className}>
    <Button onClick={resetShift}>reset</Button>
    <Button onClick={toggleEditMode} editMode={editMode}>
      <span role="img" aria-label="edit">
        ✏️
      </span>
    </Button>
    {/* sign in or out */}
  </div>
);
ControlPanel.propTypes = {
  editMode: PropTypes.bool.isRequired,
  toggleEditMode: PropTypes.func.isRequired,
  resetShift: PropTypes.func.isRequired
};

const StyledControlPanel = styled(ControlPanel)`
  position: fixed;
  z-index: 2;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 50px;
  background-color: white;
  display: flex;
`;

const Button = styled.button`
  flex-basis: 50%;
  border: 1px solid ${props => (props.editMode ? "#f5af5f" : "transparent")};
  background: ${props => (props.editMode ? "#ffe187" : "none")};
  font-size: 20px;
  font-family: monospace;
  margin: 2px;
  outline: none;

  &:active {
    border-color: #f5af5f;
    background-color: #ffe187;
  }

  &:disabled {
    opacity: 0.4;
  }
`;

export { StyledControlPanel as ControlPanel };
