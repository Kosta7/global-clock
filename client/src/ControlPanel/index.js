import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { UserLogin } from "./UserLogin";
import { UserLogout } from "./UserLogout";

export class ControlPanel extends React.Component {
  state = {
    isUserMode: false,
    userPicUrl: null
  };

  render() {
    return (
      <StyledControlPanel>
        <Button onClick={this.props.resetShift}>
          <span role="img" aria-label="reset">
            üîÉ
          </span>
        </Button>
        <Button
          onClick={this.props.toggleEditMode}
          isEditMode={this.props.isEditMode}
        >
          <span role="img" aria-label="edit">
            ‚úèÔ∏è
          </span>
        </Button>
        <Button onClick={this.toggleUserMode.bind(this)}>
          {this.props.isLoggedIn ? (
            <UserPic src={this.state.userPicUrl} />
          ) : (
            `login`
          )}
        </Button>

        {this.state.isUserMode &&
          (this.props.isLoggedIn ? (
            <UserLogout
              closeUserPopup={this.toggleUserMode.bind(this)}
              updateClockList={this.props.updateClockList.bind(this)}
            />
          ) : (
            <UserLogin
              closeUserPopup={this.toggleUserMode.bind(this)}
              updateClockList={this.props.updateClockList.bind(this)}
            />
          ))}
      </StyledControlPanel>
    );
  }

  async componentDidUpdate(prevProps) {
    if (!prevProps.isLoggedIn && this.props.isLoggedIn) {
      const response = await fetch("/auth/userpicurl");
      const userPicUrl = await response.text();
      this.setState({ userPicUrl });
    } else if (prevProps.isLoggedIn && !this.props.isLoggedIn) {
      this.setState({ userPicUrl: null });
    }
  }

  toggleUserMode() {
    this.setState({ isUserMode: !this.state.isUserMode });
  }
}
ControlPanel.propTypes = {
  isEditMode: PropTypes.bool.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  toggleEditMode: PropTypes.func.isRequired,
  resetShift: PropTypes.func.isRequired,
  updateClockList: PropTypes.func.isRequired
};

const Button = styled.button`
  flex-basis: 33%;
  border: 1px solid ${props => (props.isEditMode ? "#f5af5f" : "transparent")};
  background: ${props => (props.isEditMode ? "#ffe187" : "none")};
  font-size: 20px;
  font-family: monospace;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 2px;
  outline: none;
  cursor: pointer;

  &:active {
    border-color: #f5af5f;
    background-color: #ffe187;
  }

  &:disabled {
    opacity: 0.4;
  }
`;

const UserPic = styled.img`
  height: 30px;
  border-radius: 50%;
`;

const StyledControlPanel = styled.div`
  position: fixed;
  z-index: 2;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 50px;
  background-color: white;
  display: flex;
`;
