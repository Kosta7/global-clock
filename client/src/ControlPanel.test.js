import { ControlPanel } from "./ControlPanel";

it("renders correctly", () => {
  const wrapper = shallow(
    <ControlPanel
      incrementShift={() => {}}
      decrementShift={() => {}}
      resetShift={() => {}}
    />
  );

  expect(wrapper).toMatchSnapshot();
});

it("calls resetShift on click", () => {
  const spy = sinon.spy();
  const wrapper = mount(
    <ControlPanel
      incrementShift={() => {}}
      decrementShift={() => {}}
      resetShift={() => {}}
      toggleEditMode={spy}
    />
  );

  wrapper
    .find("button")
    .last()
    .simulate("click");

  expect(spy.calledOnce).toBe(true);
});
