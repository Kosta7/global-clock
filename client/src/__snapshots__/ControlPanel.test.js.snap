// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<StyledComponent
  decrementShift={[Function]}
  forwardedClass={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  position: fixed;
  bottom: 0;
  height: 50px;
  width: 100vw;
  display: flex;
",
        ],
      },
      "displayName": "Styled(ControlPanel)",
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
      Symbol(Symbol.hasInstance): [Function],
    }
  }
  forwardedRef={null}
  incrementShift={[Function]}
  resetShift={[Function]}
/>
`;
